pkg/controllers/maintenancelimit_controller.go

```diff
@@ -111,3 +94,89 @@ func (r *MaintenanceLimitReconciler) UpdateMaintenanceLimits(ctx context.Context
 
 	return nil
 }
+
+// getMaintenanceLimitsByType calculates the max number of nodes that can be updated at the given time
+func (r *MaintenanceLimitReconciler) getMaintenanceLimitsByType(ctx context.Context, log logr.Logger, ml *repairmanv1.MaintenanceLimit) (uint, error) {
+	if !strings.EqualFold(ml.Name, "node") {
+		return 0, errors.New("unsupported maintenance type")
+	}
+
+	var policied map[string]bool
+	err := r.getPoliciedResourcesByType(ctx, log, ml, policied)
+	if err != nil {
+		log.Error(err, "unable to get policied resource", "type", ml.Name)
+		return 0, err
+	}
+
+	var availableResources uint
+	switch ml.Name {
+	case "node":
+		var nodeList *corev1.NodeList
+		err := r.getNodes(ctx, log, nodeList)
+		if err != nil {
+			log.Error(err, "failed to list", "type", ml.Name)
+			return 0, err
+		}
+
+		if len(nodeList.Items) <= 0 {
+			return 0, nil
+		}
+		availableResources = r.calculateAvailableNodes(ctx, nodeList, policied)
+	}
+
+	nominalLimit := ml.Spec.Limit / 100 * availableResources
+	return uint(math.Max(float64(nominalLimit), float64(1))), nil
+}
+
+func (r *MaintenanceLimitReconciler) getPoliciedResourcesByType(ctx context.Context, log logr.Logger, ml *repairmanv1.MaintenanceLimit, policied map[string]bool) error {
+	switch ml.Name {
+	case "node":
+		var nodeList *corev1.NodeList
+		err := r.getNodes(ctx, log, nodeList)
+		if err != nil {
+			log.Error(err, "failed to list", "type", ml.Name)
+			return err
+		}
+
+		for _, node := range nodeList.Items {
+			r.applyNodePolicies(node, ml, policied)
+		}
+	default:
+		return errors.New("unsupported maintenance type")
+	}
+
+	return nil
+}
+
+func (r *MaintenanceLimitReconciler) applyNodePolicies(node corev1.Node, ml *repairmanv1.MaintenanceLimit, policied map[string]bool) {
+	if ml.Spec.Policies.RespectUnschedulableNodes && node.Spec.Unschedulable {
+		policied[node.Name] = true
+	}
+
+	if ml.Spec.Policies.RespectNotReadyNodes {
+		for _, condition := range node.Status.Conditions {
+			if condition.Type == corev1.NodeReady && condition.Status == corev1.ConditionFalse {
+				policied[node.Name] = true
```

Comment:
Can you create map of nodename -> node object itself

Reply:
Sounds good, will do